samza-core/src/main/java/org/apache/samza/zk/ZkMetadataStore.java

```diff
@@ -0,0 +1,117 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+import java.io.UnsupportedEncodingException;
+import java.util.concurrent.TimeUnit;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.I0Itec.zkclient.serialize.BytesPushThroughSerializer;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.ZkConfig;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.metadatastore.MetadataStore;
+import org.apache.samza.SamzaException;
+import org.I0Itec.zkclient.ZkClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of the {@link MetadataStore} interface where the
+ * metadata of the Samza job is stored in zookeeper.
+ */
+public class ZkMetadataStore implements MetadataStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZkMetadataStore.class);
+
+  private final ZkClient zkClient;
+  private final ZkConfig zkConfig;
+  private final String zkBaseDir;
+
+  public ZkMetadataStore(Config config, MetricsRegistry metricsRegistry, String zkBaseDir) {
+    this.zkConfig = new ZkConfig(config);
+    this.zkClient = new ZkClient(zkConfig.getZkConnect(), zkConfig.getZkSessionTimeoutMs(), zkConfig.getZkConnectionTimeoutMs(), new BytesPushThroughSerializer());
+    this.zkBaseDir = zkBaseDir;
+    createIfNotExists(zkBaseDir);
+  }
+
+  @Override
+  public void init() {
+    zkClient.waitUntilConnected(zkConfig.getZkConnectionTimeoutMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public byte[] get(byte[] key) {
+    return zkClient.readData(getZkPathForKey(key), true);
+  }
+
+  @Override
+  public void put(byte[] key, byte[] value) {
+    String zkPath = getZkPathForKey(key);
+    createIfNotExists(zkPath);
+    zkClient.writeData(zkPath, value);
+  }
+
+  @Override
+  public void delete(byte[] key) {
+    zkClient.delete(getZkPathForKey(key));
+  }
+
+  @Override
+  public Map<byte[], byte[]> all() {
+    try {
+      List<String> zkSubDirectories = zkClient.getChildren(zkBaseDir);
+      Map<byte[], byte[]> result = new HashMap<>();
+      for (String zkSubDir : zkSubDirectories) {
+        String completeZkPath = String.format("%s/%s", zkBaseDir, zkSubDir);
+        byte[] value = zkClient.readData(completeZkPath, true);
```

Comment:
Does this need to be done atomically for all data within the base dir?

Reply:
`ZkClient` does not have a multiGet API to read the contents of all the childNodes of a `zkRootNode` atomically. The only plausible way is to getAllChildren, iterate, read and return data for each one of them individually.  

For the use-cases we're planning to use this in samza(LeaderJobCoordinator reading/writing locality & metadata in zookeeper), I think we don't need this to be atomic.

Reply:
Current use case != all use cases, esp. for general purpose APIs like this. If this is not atomic, we should at least document the guarantee we provide in javadocs. Also, what happens if a path gets deleted while we're iterating? Does that throw an exception? Should document what this method does in that case (e.g., retries until success / ignores path in returned results / throws exception).

If we can detect missing paths, we can get a consistent result set by (bounded) retrying until we get a value for each children.

Reply:
>> Also, what happens if a path gets deleted while we're iterating? Does that throw an exception? 

`ZkClient.readData` returns null if the zookeeper path does not exists(controlled by the second parameter). 

>> Should document what this method does in that case.
Sure, would add documentation about atomicity guarantees and the above scenarios.